spot-contracts/contracts/strategies/FeeStrategy.sol

```diff
@@ -0,0 +1,226 @@
+import { IFeeStrategy, IERC20Upgradeable } from "../_interfaces/IFeeStrategy.sol";
+import { IPerpetualTranche, IBondController } from "../_interfaces/IPerpetualTranche.sol";
+import { IVault } from "../_interfaces/IVault.sol";
+
+import { MathUpgradeable } from "@openzeppelin/contracts-upgradeable/utils/math/MathUpgradeable.sol";
+import { SafeCastUpgradeable } from "@openzeppelin/contracts-upgradeable/utils/math/SafeCastUpgradeable.sol";
+import { AddressUpgradeable } from "@openzeppelin/contracts-upgradeable/utils/AddressUpgradeable.sol";
+import { OwnableUpgradeable } from "@openzeppelin/contracts-upgradeable/access/OwnableUpgradeable.sol";
+import { Sigmoid } from "../_utils/Sigmoid.sol";
+import { BondHelpers } from "../_utils/BondHelpers.sol";
+import { PerpHelpers } from "../_utils/PerpHelpers.sol";
+
+/**
+ *  @title FeeStrategy
+ *
+ *  @notice This contract determins perp's mint, burn and rollover fees.
+ *
+ *          Fees are computed based on the deviationFactor, ie the ratio between
+ *          the current TVL to the expcted TVL in the vault system.
+ *
+ *              expctedTVL = perpTVL / trancheRatio
+ *              deviationFactor = currentTVL/expectedTVL
+ *
+ *          1) The mint fees are turned on when deviationFactor < 1,
+ *             and is a fixed percentage fee set by the owner.
+ *
+ *          2) The burn fees are turned on when deviationFactor > 1,
+ *             and is a fixed percentage fee set by the owner.
+ *
+ *          3) The rollover fees are signed and can flow in either direction.
+ *             The fee is a percentage of the tranches rolled over and is computed
+ *             through a sigmoid function. The slope and asymptotes are set by the owner.
+ *
+ *              rotationsPerYear = mintingBondDuration / 1_year
+ *              rolloverFeePerc = sigmoid(deviationFactor) / rotationsPerYear
+ *
+ *
+ */
+contract FeeStrategy is IFeeStrategy, OwnableUpgradeable {
+    // Libraries
+    using MathUpgradeable for uint256;
+    using SafeCastUpgradeable for uint256;
+    using BondHelpers for IBondController;
+    using PerpHelpers for IPerpetualTranche;
+
+    /// @dev The returned fee percentages are fixed point numbers with {PERC_DECIMALS} places.
+    ///      The decimals should line up with value expected by consumer (perp).
+    uint8 public constant PERC_DECIMALS = 8;
+    uint256 public constant HUNDRED_PERC = 10**PERC_DECIMALS; // 100%
+    uint256 public constant MAX_PERC = HUNDRED_PERC / 10; // 10%
+
+    /// @dev Number of seconds in one year.
+    int256 public constant ONE_YEAR_SEC = 365 * 24 * 3600;
+
+    // Replicating value used here:
+    // https://github.com/buttonwood-protocol/tranche/blob/main/contracts/BondController.sol
+    uint256 private constant TRANCHE_RATIO_GRANULARITY = 1000;
+
+    /// @notice Reference to the perpetual token.
+    IPerpetualTranche public immutable perp;
+
+    /// @notice Reference to the rollover vault.
+    IVault public immutable vault;
```

Comment:
not necessary now, but if we ever had multiple vaults we might want this to be a list

Reply:
We could swap out the fee strategy at that point. Okay keeping it as is rn.